README.md
# Anomalous Behaviour Detection

Module responsable for detecting anomalies in case studies of algal bloom occurences based on images from Google Earth Engine API and machine learning



## Dependencies

- Python >= 3.7.7 64-bit, < 3.10 64-bit
- Google Earth Engine enabled account: see https://earthengine.google.com/


## Instalation

To install this script and all its dependencies, execute the follow commands:

1) Create a virtual environment: `python3 -m venv venv`
2) Enable it: `source venv/bin/activate`
2) Install the script dependencies: `pip install -e .`


## Attention, before running this script:

Before running the script and after installing the libraries, you must authenticate with the Google Earth Engine API using ONE of the following commands:

```bash
# from local command line
earthengine authenticate

# from inside Docker container
earthengine authenticate --auth_mode=notebook

# from inside Jupyter Notebook
import ee
ee.Authenticate()
```

In some versions of macOS, it might be necessary to run this command using `sudo`.

Additionally, make sure that folders `cache` and `results` have writing permissions:

```bash
chmod 777 /path/to/abd/cache
chmod 777 /path/to/abd/results
```

## Docker image

There is also a Docker image which provides this script with all necessary dependencies easy and ready. To use it, run:

```bash
docker run -p 8888:8888 phmananias/abd:latest
```

or you can build it locally and then run it:
```bash
docker build -t abd:latest .
docker run -p 8888:8888 abd:latest abd
```



## Command line tool

This module brings a default command line `adb` for you. To see available parameters, please run:

```bash
> abd --help
Usage: abd [OPTIONS]

Options:
  --lat_lon TEXT            Two diagnal points (Latitude 1, Longitude 1,
                            Latitude 2, Longitude 2) of the study area
  --dates TEXT              Comma-separated date to be applied the algorithm
  --name TEXT               Place where to save generated files
  --days_threshold INTEGER  Days threshold used to build the timeseries and
                            training set: 90, 180 ou 365
  --model TEXT              Select the desired model: ocsvm, rf, if or None
                            for all
  --sensor TEXT             Define the selected sensor where images will be
                            downloaded from: landsat, sentinel, modis
  --save_collection TEXT    Save collection images (tiff and png)
  --save_train TEXT         Enable saving the training dataset (csv)
  --force_cache TEXT        Force cache resetting to prevent image errors
  --attributes TEXT         Define the attributes used in the modelling
                            process
  --outliers_zscore FLOAT   Define the Z-Score used in the median outlier
                            removal threshold
  --attribute_doy TEXT      Define if the doy attribute will be used in the
                            modelling process
  --roi TEXT                Select the GEE roi version used in the validation
                            process. E.g. users/pedroananias/roi
  --cloud_threshold TEXT    Allow CLOUD threshold customization by user
  --output_folder           Specify desired results output folder
  --help                    Show this message and exit.

```

## Installing using PIP

You can directly install the latest version of this script using the following command:

```bash
pip install git+https://github.com/pedroananias/abd.git@v0.28.1
```

## How to execute the default script?

```bash
abd --lat_lon=-83.50124371805877,41.88435023280987,-83.07548096199702,41.65275061592091 --dates=2019-06-03 --name=erie --model=ocsvm --sensor=modis --output_folder=abd
```


## What are the results?

The script will detect the occurrence of anomalies with a case study of algae blooming in the inserted study area and selected date. Therefore, a folder located in `results` is created and named based on the date and version of the script executed. Example: 

```bash
/path/to/abd/results/20221023_133118[v=v0.26.0-erie,d=2019-06-03,t=180,m=ocsvm,s=modis,attr=ndvi,fai]
```

The following results are generated:

- geojson (GeoJSONs with pixels and labels)
- 2019-06-03.png (RGB image)
- 2017-12-31.zip (Bands as TIFFs files)
- detection.png (Graph of the detection result)
- results.csv (Accuracy results of validation process in CSV format -if a ROI located in GEE plataform was selected)
- timeseries.png (Graphical visualization of the built timeseries)


### Exporting GeoTIFFs to Google Drive

When using the 'save_collection' function, if the script can not save images locally, will send them to Google Drive to a folder called 'abd_name.tiff' for user who is authenticated.



## Sandbox example

This script comes with a Jupyter Notebook sandbox example file. To open it, please run the command below inside the script's root directory:

```bash
jupyter-lab
```

CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Vertex AI Pipeline implementation

## [0.27.2] - 2022-10-23
### Fixed
- Dockerfile ending after run

## [0.27.0] - 2022-10-23
### Added
- Command line tool for running detections
- Versioning control (tags, setup.py, version.py)
- Black formatting
- Docker image

### Fixed
- Python package structure best practices

## [0.26.0]
### Added
- Tests with NDVI and FAI only
- Created yearly outputs

## [0.25.0]
### Added
- SABI Index
- Isolation Forest support

## [0.24.0]
### Added
- Random Forest support for RandomizedSearchCV
- New ROIs were selected to validate the algorithm with in situ data
- New function for outlier removal
- Support for bigger study areas
- Support for Deep Learning with TensorFlow AutoEncoder

## [0.23.2]
### Added
- Built-in image caching system for increased performance

## [0.23.1]
### Fixed
- Try except when extracting pixels from images
- Trend removal correction

## [0.23.0]
### Added
- Merge test between GridSearch and RandomizedSearch
- Removal of the Outlier parameter - set to False, that is, remove
- Enabled Modis satellite

### Fixed
- Image error control, if not found, including Modis water mask
- Corrections in the generation of the recurrences graph

### Removed
- Poly kernel (optimization)

## [0.22.2]
### Added
- Modification to support mosaic of images of the same date with cut (Sentinel)

## [0.22.1]
### Added
- Creating script for cloud separation in json, correction in the display of maps

## [0.22.0]
### Added
- Trend was removed using the medium image and subtracting its pixels
from the time series images, adding the use of randomizedsearchcv

## [0.21.0-2]
### Added
- Mean and standard deviation were tested with STD, STD / 2 and STD + (STD / 2)
in versions 21.0, 21.1 and 21.2 respectively

[0.27.0]: https://github.com/pedroananias/abd/releases/tag/v0.27.0
[0.26.0]: https://github.com/pedroananias/abd/releases/tag/v0.26.0


script.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

###############################################################################################################################
# ### ABD - Anomalous Behaviour Detection
# ### Module responsable for detecting inland waters anomalies based on images from Google Earth Engine API and machine learning
# ### Python 3.7 64Bits is required!
#
# ### Change History
# - Version 21.0-2: 
# - Mean and standard deviation were tested with STD, STD / 2 and STD + (STD / 2) in versions 21.0, 21.1 and 21.2 respectively
#
# - Version 22.0: 
# - Trend was removed using the medium image and subtracting its pixels from the time series images, adding the use of randomizedsearchcv
#
# - Version 22.1: 
# - Creating script for cloud separation in json, correction in the display of maps
#
# - Version 22.2: 
# - Modification to support mosaic of images of the same date with cut (Sentinel)
#
# - Version 23.0: 
# - Merge test between GridSearch and RandomizedSearch
# - Removing the poly kernel (optimization)
# - Removal of the Outlier parameter - set to False, that is, remove
# - Enabled Modis satellite
# - added image error control, if not found, including Modis water mask
# - Corrections in the generation of the recurrences graph
#
# - Version 23.1:
# - trend removal correction
# - Added try except when extracting pixels from images
#
# - Version 23.2:
# - Built-in image caching system for increased performance
#
# - Version 24:
# - Added Random Forest support for RandomizedSearchCV
# - New ROIs were selected to validate the algorithm with in situ data
# - New function for outlier removal
# - Added support for bigger study areas
# - Added support for Deep Learning with TensorFlow AutoEncoder 
#
# - Version 25:
# - Added SABI Index
# - Added Isolation Forest support
#
# - Version 26:
# - Tests with NDVI and FAI only
# - Created yearly outputs
###############################################################################################################################

# ### Version
version = "V26"



# ### Module imports

# Main
import ee
import pandas as pd
import numpy as np
import math
import requests
import time
import warnings
import os
import sys
import argparse
import logging
import traceback
import itertools
import gc

# Sub
from datetime import datetime as dt
from datetime import timedelta

# Extras modules
from modules import misc, gee, abd

# ignore warnings
warnings.filterwarnings('ignore')


# ### Script args parsing

# starting arg parser
parser = argparse.ArgumentParser(description=version)

# create arguments
parser.add_argument('--lat_lon', dest='lat_lon', action='store', default="-83.50124371805877,41.88435023280987,-83.07548096199702,41.65275061592091",
                   help="Two diagnal points (Latitude 1, Longitude 1, Latitude 2, Longitude 2) of the study area")
parser.add_argument('--dates', dest='dates', action='store', default="2019-06-03,2019-07-01,2019-08-19,2019-09-24",
                   help="Comma-separated date to be applied the algorithm")
parser.add_argument('--name', dest='name', action='store', default="erie",
                   help="Place where to save generated files")
parser.add_argument('--days_threshold', dest='days_threshold', action='store', type=int, default=180,
                   help="Days threshold used to build the timeseries and training set: 90, 180 ou 365")
parser.add_argument('--model', dest='model', action='store', default=None,
                   help="Select the desired module: ocsvm, rf, ae, if or None for all")
parser.add_argument('--sensor', dest='sensor', action='store', default="modis",
                   help="Define the selected sensor where images will be downloaded from: landsat, sentinel, modis")
parser.add_argument('--save_collection', dest='save_collection', action='store_true',
                   help="Save collection images (tiff and png)")
parser.add_argument('--save_train', dest='save_train', action='store_true',
                   help="Enable saving the training dataset (csv)")
parser.add_argument('--force_cache', dest='force_cache', action='store_true',
                   help="Force cache reseting to prevent image errors")
parser.add_argument('--attributes', dest='attributes', action='store', default="ndvi,fai",
                   help="Define the attributes used in the modelling process")
parser.add_argument('--outliers_zscore', dest='outliers_zscore', action='store', type=float, default=3.0,
                   help="Define the Z-Score used in the median outlier removal threshold")
parser.add_argument('--attribute_doy', dest='attribute_doy', action='store_true',
                   help="Define if the doy attribute will be used in the modelling process")
parser.add_argument('--roi', dest='roi', action='store', default="",
                   help="Select the roi version used in the validation process")
parser.add_argument('--cloud_threshold', dest='cloud_threshold', action='store', type=float, default=None,
                   help="Allow CLOUD threshold customization by user")

# parsing arguments
args = parser.parse_args()


# ### Start

try:

  # Start script time counter
  start_time = time.time()

  # Google Earth Engine API initialization
  ee.Initialize()



  # ### Working directory

  # Data path
  folderRoot = os.path.dirname(os.path.realpath(__file__))+'/data'
  if not os.path.exists(folderRoot):
    os.mkdir(folderRoot)

  # Images path
  folderCache = os.path.dirname(os.path.realpath(__file__))+'/cache'
  if not os.path.exists(folderCache):
    os.mkdir(folderCache)



  # ### ABD execution

  # check if date range was selected and build time series based on it
  yearly = False
  if "*" in args.dates:
    yearly = True
    path_df_yearly_occurrence = folderRoot+'/df_yearly_occurrence.csv'
    start = dt.strptime(args.dates.split("*")[0], "%Y-%m-%d")
    end = dt.strptime(args.dates.split("*")[1], "%Y-%m-%d")
    dates = [(start + timedelta(days=x)).strftime("%Y-%m-%d") for x in range(0, (end-start).days+1)]
    args.dates = ",".join(dates)

  # go through each parameter
  for date in args.dates.split(","):

    # folder to save results from algorithm at
    folder = folderRoot+'/'+dt.now().strftime("%Y%m%d_%H%M%S")+'[v='+str(version)+'-'+str(args.name)+',d='+str(date)+',t='+str(args.days_threshold)+',m='+str(args.model)+',s='+str(args.sensor)+',attr='+str(args.attributes)+']'
    if not os.path.exists(folder):
      os.mkdir(folder)

    # create ABD algorithm
    algorithm = abd.Abd(lat_lon=args.lat_lon,
                        date=date,
                        days_threshold=args.days_threshold,
                        model=args.model,
                        sensor=args.sensor,
                        scaler='robust',
                        cache_path=folderCache,
                        force_cache=args.force_cache,
                        remove_outliers=True,
                        reduce_dimensionality=False,
                        morph_op=None,
                        morph_op_iters=1,
                        convolve=False,
                        convolve_radius=1,
                        outliers_zscore=args.outliers_zscore,
                        attributes=args.attributes.split(","),
                        attribute_doy=args.attribute_doy,
                        cloud_threshold=args.cloud_threshold)

    # preprocessing
    algorithm.process_timeseries_data()

    # creating train and grid datasets
    algorithm.process_training_data(df=algorithm.df_timeseries)

    # start training process
    algorithm.train()

    # apply detection algorithm
    algorithm.detect()

    # validate using ROI
    algorithm.validate_using_roi(path='users/pedroananias/'+str(args.name), rois=[date+'_'+args.sensor+'_regular'+args.roi, date+'_'+args.sensor+'_anomaly'+args.roi], labels=[0, 1])

    # save geojson
    algorithm.save_geojsons(folder=folder+"/geojson")

    # save results
    algorithm.save_dataset(df=algorithm.df_results, path=folder+'/results.csv')

    # save collection images
    if isinstance(algorithm.df_image, pd.DataFrame) and not algorithm.df_image.empty:
      algorithm.save_detection_plot(path=folder+'/detection.png')
      algorithm.save_image_png(image=algorithm.image_clip, date=dt.strptime(date, "%Y-%m-%d"), path=folder+"/"+date+".png")
      algorithm.save_image_tiff(image=algorithm.image_clip, date=dt.strptime(date, "%Y-%m-%d"), path=folder+"/"+date+".zip", folderName=args.name)
      
    # save collection images (tiff and png)
    if args.save_collection:
      algorithm.save_timeseries_plot(df=algorithm.df_timeseries, path=folder+'/timeseries.png')
      algorithm.save_collection_tiff(folder=folder+"/tiff", folderName=args.name)
      algorithm.save_collection_png(folder=folder+"/png")

    # save preprocessing results
    if args.save_train:
      algorithm.save_dataset(df=algorithm.df_timeseries, path=folder+'/timeseries.csv')
      algorithm.save_dataset(df=algorithm.df_train, path=folder+'/df_train.csv')
      algorithm.save_dataset(df=algorithm.df_gridsearch, path=folder+'/df_gridsearch.csv')
      algorithm.save_dataset(df=algorithm.df_image, path=folder+'/df_image.csv')

    # results
    # add results and save it on disk
    if not algorithm.df_results is None and not algorithm.df_results.empty:
      path_df_results = folderRoot+'/results.csv'
      df_results = pd.read_csv(path_df_results).drop(['Unnamed: 0'], axis=1, errors="ignore").append(algorithm.df_results) if os.path.exists(path_df_results) else algorithm.df_results.copy(deep=True)
      df_results.to_csv(r''+path_df_results)

    # yearly occurrence
    # add results to yearly occurrence
    if yearly and isinstance(algorithm.df_image, pd.DataFrame) and not algorithm.df_image.empty:
      df_image = algorithm.df_image.astype(str)
      df_yearly_occurrence = pd.read_csv(path_df_yearly_occurrence).drop(['Unnamed: 0'], axis=1, errors="ignore").append(df_image) if os.path.exists(path_df_yearly_occurrence) else df_image.copy(deep=True)
      df_yearly_occurrence.to_csv(r''+path_df_yearly_occurrence)

      # clear memory
      del df_yearly_occurrence
      gc.collect()

    # clear memory
    del algorithm
    gc.collect()

  # check if it should build yearly occurrences and save it
  if yearly and os.path.exists(path_df_yearly_occurrence):

    # fix/build yearly dataframe
    df_yearly_occurrence = pd.read_csv(path_df_yearly_occurrence).drop(['Unnamed: 0','pixel','index','doy','ndvi','fai','ndvi_median','fai_median','ndvi_diff','fai_diff'], axis=1, errors="ignore")
    df_yearly_occurrence.rename(columns={"label_0": "occurrence"}, inplace=True)
    df_yearly_occurrence['year'] = pd.to_datetime(df_yearly_occurrence['date'], format='%Y-%m-%d', errors='ignore').dt.year.astype(dtype=np.int64, errors='ignore')
    df_yearly_occurrence['cloud'] = df_yearly_occurrence['cloud'].astype(dtype=np.int64, errors='ignore')
    df_yearly_occurrence['not_occurrence'] = 0

    # sum all occurrences, clouds and non_occurrence by year, lat and lon
    df_yearly_occurrence.loc[((df_yearly_occurrence['occurrence']==0.0) & (df_yearly_occurrence['cloud']==0.0)), 'not_occurrence'] = 1
    df_yearly_occurrence = df_yearly_occurrence.groupby(['lat','lon','year']).sum().reset_index()

    # add porcentage of occurrence and cloud
    df_yearly_occurrence['pct_occurrence']   = (df_yearly_occurrence['occurrence']/(df_yearly_occurrence['occurrence']+df_yearly_occurrence['not_occurrence']))*100
    df_yearly_occurrence['pct_cloud']        = (df_yearly_occurrence['cloud']/(df_yearly_occurrence['occurrence']+df_yearly_occurrence['not_occurrence']+df_yearly_occurrence['cloud']))*100
    df_yearly_occurrence['instants']         = df_yearly_occurrence['occurrence']+df_yearly_occurrence['not_occurrence']+df_yearly_occurrence['cloud']

    # fix problem with 1 instant pixels for series average bigger than 10 instants
    if df_yearly_occurrence['instants'].mean()>10:
      df_yearly_occurrence.loc[(df_yearly_occurrence['instants']==1), 'pct_occurrence'] = np.nan
      df_yearly_occurrence['pct_occurrence'].fillna(method='ffill', inplace=True)

    # folder to save results from algorithm at
    folder = folderRoot+'/'+dt.now().strftime("%Y%m%d_%H%M%S")+'[v='+str(version)+'-'+str(args.name)+']'
    if not os.path.exists(folder):
      os.mkdir(folder)

    # create new ABD intance
    algorithm = abd.Abd(lat_lon=args.lat_lon,
                        date=args.dates.split(",")[0],
                        days_threshold=args.days_threshold,
                        model=args.model,
                        sensor=args.sensor,
                        scaler='robust',
                        cache_path=folderCache,
                        force_cache=args.force_cache,
                        remove_outliers=True,
                        reduce_dimensionality=False,
                        morph_op=None,
                        morph_op_iters=1,
                        convolve=False,
                        convolve_radius=1,
                        outliers_zscore=args.outliers_zscore,
                        attributes=args.attributes.split(","),
                        attribute_doy=args.attribute_doy,
                        cloud_threshold=args.cloud_threshold)

    # save occurrences plot
    algorithm.save_occurrences_plot(df=df_yearly_occurrence, path=folder+'/yearly_svm.png')

    # save occurrences geojson
    algorithm.save_occurrences_geojson(df=df_yearly_occurrence, path=folder+'/yearly_svm.json')

  # ### Script termination notice
  script_time_all = time.time() - start_time
  debug = "***** Script execution completed successfully (-- %s seconds --) *****" %(script_time_all)
  print()
  print(debug)

except:

    # ### Script execution error warning

    # Execution
    print()
    print()
    debug = "***** Error on script execution: "+str(traceback.format_exc())
    print(debug)

    # Removes the folder created initially with the result of execution
    script_time_all = time.time() - start_time
    debug = "***** Script execution could not be completed (-- %s seconds --) *****" %(script_time_all)
    print(debug)